src/NServiceBus.Core/OpenTelemetry/Metrics/MessageHandlingMetricsFactory.cs

```diff
@@ -0,0 +1,43 @@
+﻿namespace NServiceBus;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Pipeline;
+
+class MessageHandlingMetricsFactory(string queueName, string discriminator) : IMessageHandlingMetricsFactory
+{
+    public IMessageHandlingMetrics StartHandling(IInvokeHandlerContext context)
+    {
+        var messageType = context.MessageBeingHandled?.GetType().FullName;
+        TagList tagList = MeterTags.CommonMessagingMetricTags(queueName, discriminator, messageType);
+        var handlerType = context.MessageHandler.Instance?.GetType().FullName;
+        tagList.Add(MeterTags.MessageHandlerType, handlerType ?? "");
+        return new RecordMessageHandlingMetric(tagList);
+    }
+}
+
+class RecordMessageHandlingMetric : IMessageHandlingMetrics
+{
+    public RecordMessageHandlingMetric(TagList tags)
+    {
+        this.tags = tags;
+        stopWatch.Start();
+    }
+
+    public void OnSuccess()
+    {
+        stopWatch.Stop();
```

Comment:
Do we want to add the execution results? This would allow us to remove the existing counters in the future (`TotalProcessedSuccessfully` and `TotalFailures`).

```suggestion
        stopWatch.Stop();
        tags.Add(new KeyValuePair<string, object>("execution.result", "success"));
```